### PR TITLE
contractcourt: consider delivery addresses when evaluating toSelfAmount

### PR DESCRIPTION
**Pull request by ProofOfKeags**
## Change Description
This commit fixes https://github.com/lightningnetwork/lnd/issues/8535 by changing how we assess toSelfAmount inside
the chainWatcher.

In certain cases users may wish to close out channel funds to external
delivery addresses set either during open or close.

Prior to this change we only consider addresses that our wallet is
aware of and permitted the possibility of multiple to_self outputs,
which is impossible according to the protocol.

This change now identifies to_self outputs based off of matching
delivery scripts, and if no delivery script is set, fall back to
the first output that our wallet can recognize.

NOTE FOR REVIEWERS: The itest commit is ordered first to confirm that the test fails prior to the change. After the fix commit, the test passes. If you want to confirm that the test is correctly designed you can run the test prior to the fix commit.

## Steps to Test
`make itest icase=coop_close_with_external_delivery`

### Testing
- [ ] Your PR passes all CI checks.
- [ ] Tests covering the positive and negative (error paths) are included.
- [ ] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [ ] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [ ] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [ ] Any new logging statements use an appropriate subsystem and logging level.
- [ ] Any new lncli commands have appropriate tags in the comments for the rpc in the proto file.
- [ ]  [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our  [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

### CODE REVIEW

File: itest/lnd_coop_close_external_delivery_test.go
Comment (ziggie1984) [inline itest/lnd_coop_close_external_delivery_test.go:33]: So for me the test passes without funding Alice with extra coins, commented this line out and run the test, everything works fine

File: itest/lnd_coop_close_external_delivery_test.go
Comment (ziggie1984) [inline itest/lnd_coop_close_external_delivery_test.go:62]: Nit: This is already done in `ht.OpenChannel`

File: itest/lnd_coop_close_external_delivery_test.go
Comment (ziggie1984) [inline itest/lnd_coop_close_external_delivery_test.go:13]: Nit: set at open => set addr at open
set at close => set addr at close

File: contractcourt/chain_watcher.go
Comment (ziggie1984) [inline contractcourt/chain_watcher.go:993]: Nit: chainWatcher's channel

File: contractcourt/chain_watcher.go
Comment (ziggie1984) [inline contractcourt/chain_watcher.go:1014]: I think it makes sense to either introduce this new way of commenting function into the coding docs so that we don't diverge in our coding style ?

File: contractcourt/chain_watcher.go
Comment (ziggie1984) [inline contractcourt/chain_watcher.go:1017]: Nit: explanation why it does not matter would be helpful.

File: contractcourt/chain_watcher.go
Comment (ziggie1984) [inline contractcourt/chain_watcher.go:1011]: I am not sure I understand this comment, seems to me we currently only use this function in the cooperate close case, why do we mention the force closure output here ?

**Comment by ziggie1984** [comment]
https://github.com/lightningnetwork/lnd/pull/8808 was merged so this can be updated now.

**Comment by Crypt-iQ** [comment]
think gomod needs to be updated to compile. linter failing as well. code looks good can check again when these are fixed

File: contractcourt/chain_watcher.go
Comment (ProofOfKeags) [reply contractcourt/chain_watcher.go:1011]: While I hadn't looked at this prior to you pointing this out, it *does* appear that this function's only use site is in a context where we already assume/know it to be a coop close. That said, given that the function is named `toSelfAmount`, I think it's fair to keep it general enough to handle either case.

Generally, since the logical dependency relationship is that the caller depends on the callee, I don't think it's reasonable to assume how a function might be called. Perhaps if we want to limit it I can rename it to `coopToSelfAmount` and then trim out any commentary about how it might be used in a FC case. Perhaps this is the right compromise.

File: contractcourt/chain_watcher.go
Comment (ProofOfKeags) [reply contractcourt/chain_watcher.go:1014]: I'm not sure what the specific ask is here.

The annotation I have here is for informational purposes only to alleviate the need to squint at the function itself to figure out its "shape". The format I use is stolen from Haskell and subtle variations can be found in Standard ML, OCaml, Haskell, Idris, Agda, F#, etc.

In recent years it has been adapted into other more mainstream languages in documentation notes like you can see [here](https://github.com/fantasyland/fantasy-land/tree/v4.0.1), [here](https://github.com/ramda/ramda/wiki/Type-Signatures#named-types) and [here](https://github.com/sanctuary-js/sanctuary). The reason for this is that it's a fairly space efficient way to communicate the shape of the function without diving into the implementation details.

Are you saying that we need a note in the style guidelines to bless this? Or is it to describe in detail how to interpret a standalone type signature? What would you like to see?

File: contractcourt/chain_watcher.go
Comment (ProofOfKeags) [reply contractcourt/chain_watcher.go:1017]: I wish I knew 😅. I've restored the comment to its original text now. Also worth noting that this is inside the txscript package in btcd.

File: itest/lnd_coop_close_external_delivery_test.go
Comment (ProofOfKeags) [reply itest/lnd_coop_close_external_delivery_test.go]: It magically resolved itself on my machine too... 😒

File: contractcourt/chain_watcher.go
Comment (ziggie1984) [reply contractcourt/chain_watcher.go:1014]: > Are you saying that we need a note in the style guidelines to bless this?

Yes this is what I meant, because I find it useful.

**Review by ziggie1984 — APPROVED** [review]
LGTM ⭐️

File: contractcourt/chain_watcher.go
Comment (Chinwendu20) [inline contractcourt/chain_watcher.go:984]: Why does this end with a colon?

File: contractcourt/chain_watcher.go
Comment (Chinwendu20) [inline contractcourt/chain_watcher.go:985]: so we do not handle DB error, here?

File: contractcourt/chain_watcher.go
Comment (Chinwendu20) [inline contractcourt/chain_watcher.go:979]:
```go
// toSelfAmount takes a transaction and returns the sum of all outputs that pay
// to a script that the wallet controls. If no outputs pay to us, then we
// return zero. This is possible as our output may have been trimmed due to
// being dust.
```

This is the current godoc of this function, I think we need to update it as a result of this change

File: contractcourt/chain_watcher.go
Comment (Chinwendu20) [inline contractcourt/chain_watcher.go:991]: I do not think this function picks out the TxOut per se, but just checks?

File: itest/lnd_coop_close_external_delivery_test.go
Comment (Chinwendu20) [inline itest/lnd_coop_close_external_delivery_test.go:21]: Interesting choice, why this and not TDD?

**Review by Chinwendu20 — COMMENTED** [review]
Nice use of the fn package in this PR 😎

File: itest/lnd_coop_close_external_delivery_test.go
Comment (Chinwendu20) [inline itest/lnd_coop_close_external_delivery_test.go:20]: Interesting choice, why this and not TDD?

File: contractcourt/chain_watcher.go
Comment (ProofOfKeags) [reply contractcourt/chain_watcher.go:984]: to indicate that the immediately following code implements the stated logic.

File: contractcourt/chain_watcher.go
Comment (ProofOfKeags) [reply contractcourt/chain_watcher.go:985]: Since this function doesn't return an error, if we can't reach the shutdown info for whatever reason we treat it as non-existent. Why it failed is inconsequential.

File: contractcourt/chain_watcher.go
Comment (ProofOfKeags) [reply contractcourt/chain_watcher.go:979]: Good catch.

File: contractcourt/chain_watcher.go
Comment (ProofOfKeags) [reply contractcourt/chain_watcher.go:991]: Yeah good call this is leftover from a previous draft. Will update.

File: itest/lnd_coop_close_external_delivery_test.go
Comment (ProofOfKeags) [reply itest/lnd_coop_close_external_delivery_test.go:20]: What do you mean by TDD? I'm only familiar with the acronym as Test Driven Development, and that is exactly what I've done here.

Comment (Chinwendu20) [reply itest/lnd_coop_close_external_delivery_test.go:20]: Oh, I was referring to something like this:
image

source: https://golang.cafe/blog/golang-table-test-example

In the code being reviewed above, a number of things were rewritten twice.

File: contractcourt/chain_watcher.go
Comment (Chinwendu20) [reply contractcourt/chain_watcher.go:986]: Oh okay, I was just thinking that maybe if there is something wrong with the DB we can catch it as soon as we encounter it here.

Comment (ProofOfKeags) [reply contractcourt/chain_watcher.go:986]: If this function already had an error in its signature then I'd probably do that. But this would cause a currently non-erroring function to return an error which I'd like to avoid. Swallowing errors *can* be dangerous but in this case all it does is affect the numbers we report and I don't think that's especially problematic. If it comes out to result in bad UX's we can come back and address it but for now I'd recommend leaving it as is.

File: itest/lnd_coop_close_external_delivery_test.go
Comment (ProofOfKeags) [reply itest/lnd_coop_close_external_delivery_test.go:20]: I stole the test vector approach from @ellemouton 's refactoring of one of my earlier itests here:

https://github.com/lightningnetwork/lnd/blob/master/itest/lnd_coop_close_with_htlcs_test.go#L24-L34

I also recall @yyforyongyu saying that on a go-forward basis we should discourage table tests like in the image you linked. I tend to agree with this assessment but also would defer this type of decision to the elder wizards. I don't have a principled opinion on how to code test vectors.

**Review by ViktorTigerstrom — APPROVED** [review]
LGTM! Really elegant solution, great work 🚀!

Nit:
If anything, I think the test coverage could be extended a bit to improve the coverage further. I'm therefore leaving some suggestions below on how the coverage could be improved. None of this is blocking, and I'm fine with leaving it as is, so I leave it up to you to decide if you feel it's worth adding:

1. Unless I'm missing something, we're currently missing any itest coverage for `SettledBalance` in general. I therefore think it would make sense to rename the integration test to `testChannelSettledBalance` (since we currently have a `testChannelUnsettledBalance` integration test) and also include coverage for cases where no `DeliveryAddress` has been specified.
2. Instead of only asserting that the channel's `SettledBalance` is `NotZero` at the last line of `testCoopCloseWithExternalDeliveryImpl`, I think it would make sense to assert that the `SettledBalance` is the actual value we expect. It would then also make sense to add a `PushAmt` value to the `OpenChannelParams`.
3. Furthermore, it would be great if you could add test coverage that proves that the `toSelfAmount` returns the sum of all outputs for which the address either matches the delivery script or is an address that our wallet is aware of. To provide that coverage, after adding a `PushAmt` value to the `OpenChannelParams`, you could potentially ensure that Alice sets the `CloseAddress` to Bob's address in the `OpenChannelParams`, but then let Bob initiate the `CloseChannel` RPC with one of Alice's addresses set as the `DeliveryAddress` in the `CloseChannelRequest`.

Like I said, these are non-blocking suggestions. Great work 🎉!

**Comment by ProofOfKeags** [comment]
> If anything, I think the test coverage could be extended a bit to improve the coverage further. I'm therefore leaving some suggestions below on how the coverage could be improved. None of this is blocking, and I'm fine with leaving it as is, so I leave it up to you to decide if you feel it's worth adding:

I'm amenable to this.

> Unless I'm missing something, we're currently missing any itest coverage for SettledBalance in general. I therefore think it would make sense to rename the integration test to testChannelSettledBalance (since we currently have a testChannelUnsettledBalance integration test) and also include coverage for cases where no DeliveryAddress has been specified.

Got it. I can add the case where it isn't specified.

> Instead of only asserting that the channel's SettledBalance is NotZero at the last line of testCoopCloseWithExternalDeliveryImpl, I think it would make sense to assert that the SettledBalance is the actual value we expect. It would then also make sense to add a PushAmt value to the OpenChannelParams

This is a bit tricky as determining the balances of coop close transactions are subject to fee and output script factors and this can make solving for the "expected" amount equivalent to solving for the actual amount. We could use that PushAmt though to assert just that it's less/greater than some amount.

> Furthermore, it would be great if you could add test coverage that proves that the toSelfAmount returns the sum of all outputs for which the address either matches the delivery script or is an address that our wallet is aware of. To provide that coverage, after adding a PushAmt value to the OpenChannelParams, you could potentially ensure that Alice sets the CloseAddress to Bob's address in the OpenChannelParams, but then let Bob initiate the CloseChannel RPC with one of Alice's addresses set as the DeliveryAddress in the CloseChannelRequest.

This is the suggestion I understand the least. I think the property that you are trying to ensure in this case is that if they each set their delivery address to one that the other controls, then their settled balances should always match each other, irrespective of what the push amount is. Am I interpreting this right?

File: itest/lnd_coop_close_external_delivery_test.go
Comment (yyforyongyu) [reply itest/lnd_coop_close_external_delivery_test.go:20]: Just my two cents,

For this case, I would do `testCoopCloseWithExternalDeliveryImpl` and `testCoopCloseWithExternalDeliveryImplWithUpfronShutdown` instead.

I like using table-driven tests when the outputs of a function are only determined by one or two inputs in a test setup, like this [one](https://github.com/lightningnetwork/lnd/blob/master/buffer/buffer_test.go#L35) or this [one](https://github.com/lightningnetwork/lnd/blob/master/lnwire/writer_test.go#L464). They share common heuristics like,
- the method being tested is relatively simple - simple as in the number of inputs it takes and the number of outputs it returns, lines of code, logic, etc.
- the test itself is straightforward, which is usually not the case in itest as we often need a somewhat complicated setup to reach one or multiple states before the actual test.
- the assertions used in the test are **NOT** conditional based on the values of the table. An example I want to discourage is the [multi-hop itest](https://github.com/lightningnetwork/lnd/blob/master/itest/lnd_multi-hop_test.go) it took me quite some time to maintain while working on the sweeper series and I'm refactoring them now.

I'm also struggling to apply these rules of thumb from time to time because we need to improve the testability of the code. My general take is, as long as the tests are easy to maintain I'm happy. This means if table-driven tests are not used, I'm good, as the tests can always be refactored to use tables (low priority IMO); if table-driven tests are **properly** used, I'm good. I'm against improper usage of table-driven tests which lead to high maintenance costs.

So in **most** of cases, I'd avoid using table-driven tests in itest. I'd also recommend this [Advanced Testing with Go](https://www.youtube.com/watch?v=8hQG7QlcLBk&t=1234s&ab_channel=GopherAcademy) talk, especially the "REPEAT YOURSELF" part.

Finally I would be very hesitant to make changes to any PR, unless critical, if it already got two approvals and no Request-for-change from others🤓

**Comment by ViktorTigerstrom** [comment]
> I'm amenable to this.

Awesome! 

> Got it. I can add the case where it isn't specified.

👍 

> This is a bit tricky as determining the balances of coop close transactions are subject to fee and output script factors and this can make solving for the "expected" amount equivalent to solving for the actual amount. We could use that PushAmt though to assert just that it's less/greater than some amount.

Wouldn't we be able to just assert that the `SettledBalance` matches the sum of the `TxOuts` of the closing tx (returned by `ht.MineClosingTx(chanPoint)`) that we're interested in for the respective the test? If you feel that this get's overcomplicated though, feel free to implement your `PushAmt` less/greater idea though!

> This is the suggestion I understand the least. I think the property that you are trying to ensure in this case is that if they each set their delivery address to one that the other controls, then their settled balances should always match each other, irrespective of what the push amount is. Am I interpreting this right?

Sorry, maybe I was a bit too unclear! I hope this is clearer, but let me know if I'm still too unclear:
What I was referring to was that the current test code only covers a case where only the `isDeliveryOutput` function matches in the `toSelfAmount` function. If you implement suggestion (1.) of my previous comment, that should add coverage for a case where only the `isWalletOutput` function matches.
What my suggestion (3.) is trying to achieve, is to implement coverage for the edge case where there are outputs in the closing tx where **_both_** the `isDeliveryOutput` function and the `isWalletOutput` matches, just to prove that the function indeed sums the amount of those outputs if that's the case. My suggestion was attempting to achieve such a case in a simple way, as Alice set `DeliveryAddress` will then make the `isDeliveryOutput` function match, and Bob's set `DeliveryAddress` will then make the `isWalletOutput` function match when evaluating the `SettledBalance` from Alice perspective.
Though, if you have an easier way of achieving a scenario where there are outputs in the closing tx where **_both_** the `isDeliveryOutput` function and the `isWalletOutput` matches, feel more than free to implement that!

**Review by Crypt-iQ — APPROVED** [review]

**Pull request merged**
